---
title: "Working with larger VCF: T2T by chromosome"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Working with larger VCF: T2T by chromosome}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Overview

In this document we illustrate some issues with
large data volumes.

Here is how we acquire the genotypes for the thousand
genomes samples based on the T2T reference.  We obtained
bgzipped vcf via
```
AnVIL::gsutil_cp("gs://fc-47de7dae-e8e6-429c-b760-b4ba49136eee/1KGP/joint_genotyping/joint_vcfs/raw/chr22.genotyped.vcf.gz", ".")
```
Then we used hail in python to deal with the conversion to MatrixTable.
We could have done this in R, but we had to learn how to manipulate
the 'reference sequence'.  We have a conjectural 'reference sequence'
json document in the json folder of the BiocHail package, used here
as `t2tAnVIL.json`.
```
>>> import hail as h
>>> rg = h.get_reference('GRCh38')
Initializing Hail with default parameters...
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Running on Apache Spark version 3.1.3
SparkUI available at http://756809c79837:4040
Welcome to
     __  __     <>__
    / /_/ /__  __/ /
   / __  / _ `/ / /
  /_/ /_/\_,_/_/_/   version 0.2.105-acd89e80c345
LOGGING: writing to /home/rstudio/hail-20221213-1558-0.2.105-acd89e80c345.log
>>> nn = rg.read("t2tAnVIL.json")
>>> h.import_vcf("chr22.genotyped.vcf.gz", force_bgz=True, reference_genome=nn).write("t2t22.mt")
```
This operation seems to take a long time even with 64 cores.  We
could not get exact timing owing to some connectivity problems.

```{r doini,message=FALSE}
library(BiocHail)
```
```{r do22}
hl <- hail_init()
nn <- hl$get_reference('GRCh38')
nn <- nn$read(system.file("json/t2tAnVIL.json", package="BiocHail"))
mt22 <- hl$read_matrix_table("t2t22.mt")
mt22$count()
```
